File: problem-solving/adventofcode2020/day4.py
# https://adventofcode.com/2020/day/4
import re

input_file = open('./day4.txt', 'r')
input_list = input_file.read().splitlines()

def solve_first():
    return get_valid_count(input_list, has_all_required_fields)

def solve_second():
    return get_valid_count(input_list, has_all_required_fields_and_valid)

def get_valid_count(input_list, check_valid_func):
    passport_data = {}
    valid_count = 0
    for input in input_list:
        if input == "":
            if check_valid_func(passport_data):
                valid_count += 1
            passport_data = {}
            continue
        key_values = input.split(' ')
        for key_value in key_values:
            key, value = key_value.split(':')
            passport_data[key] = value
    if check_valid_func(passport_data):
        valid_count += 1
    return valid_count

def has_all_required_fields(passport_data):
    return all(field in passport_data for field in ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"])

def has_all_required_fields_and_valid(p):
    return check_byr(p) and check_iyr(p) and check_eyr(p) and check_hgt(p) and check_hcl(p) and check_ecl(p) and check_pid(p)

def check_byr(passport_data):
    if "byr" not in passport_data:
        return False
    byr = passport_data["byr"]
    if not byr.isdigit() or len(byr) != 4:
        return False
    return 1920 <= int(byr) <= 2002

def check_iyr(passport_data):
    if "iyr" not in passport_data:
        return False
    iyr = passport_data["iyr"]
    if not iyr.isdigit() or len(iyr) != 4:
        return False
    return 2010 <= int(iyr) <= 2020

def check_eyr(passport_data):
    if "eyr" not in passport_data:
        return False
    eyr = passport_data["eyr"]
    if not eyr.isdigit() or len(eyr) != 4:
        return False
    return 2020 <= int(eyr) <= 2030

def check_hgt(passport_data):
    if "hgt" not in passport_data:
        return False
    hgt = passport_data["hgt"]
    followed = hgt[-2:]
    if followed == "cm":
        if not 150 <= int(hgt[:-2]) <= 193:
            return False
    elif followed == "in":
        if not 59 <= int(hgt[:-2]) <= 76:
            return False
    else:
        return False
    return True

def check_hcl(passport_data):
    if "hcl" not in passport_data:
        return False
    hcl = passport_data["hcl"]
    if len(hcl) != 7:
        return False
    if hcl[0] != "#":
        return False
    p = re.compile("[0-9a-f]+")
    return p.match(hcl[1:])

def check_ecl(passport_data):
    if "ecl" not in passport_data:
        return False
    return passport_data["ecl"] in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]

def check_pid(passport_data):
    if "pid" not in passport_data:
        return False
    pid = passport_data["pid"]
    return pid.isdigit() and len(pid) == 9

if __name__ == "__main__":
    print(solve_first())
    print(solve_second())

File: problem-solving/adventofcode2020/day4.cpp
// https://adventofcode.com/2020/day/4
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <iterator>
#include <map>
#include <regex>

using namespace std;

vector<map<string, string> > parseInput(string filepath)
{
    vector<map<string, string> > passports;
    ifstream inputFile;
    inputFile.open(filepath);

    map<string, string> passport;
    string line;
    const regex reg("([a-z]{3}):([^\\s]+)");
    while (getline(inputFile, line))
    {
        sregex_token_iterator iter(line.begin(), line.end(), reg, {1, 2});
        while (iter != sregex_token_iterator())
        {
            string key = *iter;
            iter = next(iter);
            string val = *iter;
            passport[key] = val;
            iter = next(iter);
        }

        if (line == "")
        {
            passports.push_back(passport);
            passport.clear();
            continue;
        }
    }
    passports.push_back(passport);
    passport.clear();
    inputFile.close();
    return passports;
}

bool validateByr(map<string, string> &passport)
{
    if (passport["byr"] == "")
        return false;
    int byr = stoi(passport["byr"]);
    if (byr >= 1920 && byr <= 2002)
        return true;
    return false;
}

bool validateIyr(map<string, string> &passport)
{
    if (passport["iyr"] == "")
        return false;
    int iyr = stoi(passport["iyr"]);
    if (iyr >= 2010 && iyr <= 2020)
        return true;
    return false;
}

bool validateEyr(map<string, string> &passport)
{
    if (passport["eyr"] == "")
        return false;
    int eyr = stoi(passport["eyr"]);
    if (eyr >= 2020 && eyr <= 2030)
        return true;
    return false;
}

bool validateHgt(map<string, string> &passport)
{
    if (passport["hgt"] == "")
        return false;
    const regex cmRegex("(^[0-9]{3}cm$)");
    const regex inRegex("(^[0-9]{2}in$)");

    sregex_token_iterator iter(passport["hgt"].begin(), passport["hgt"].end(), cmRegex, 1);
    if (iter != sregex_token_iterator())
    {
        int hgt = stoi(*iter);
        if (hgt >= 150 && hgt <= 193)
        {
            return true;
        }
    }

    iter = {passport["hgt"].begin(), passport["hgt"].end(), inRegex, 1};
    if (iter != sregex_token_iterator())
    {
        int hgt = stoi(*iter);
        if (hgt >= 59 && hgt <= 76)
            return true;
    }
    return false;
}

bool validateHcl(map<string, string> &passport)
{
    if (passport["hcl"] == "")
        return false;
    const regex reg("(^#[0-9a-f]{6}$)");
    sregex_token_iterator iter(passport["hcl"].begin(), passport["hcl"].end(), reg, 1);
    if (iter != sregex_token_iterator())
        return true;
    return false;
}

bool validateEcl(map<string, string> &passport)
{
    if (passport["ecl"] == "")
        return false;
    const vector<string> ecls{"amb", "blu", "brn", "gry", "grn", "hzl", "oth"};
    for (const string &ecl : ecls)
        if (passport["ecl"] == ecl)
            return true;
    return false;
}

bool validatePid(map<string, string> &passport)
{
    if (passport["pid"] == "")
        return false;
    const regex reg("^[0-9]{9}$");
    sregex_token_iterator iter(passport["pid"].begin(), passport["pid"].end(), reg, 1);
    if (iter != sregex_token_iterator())
        return true;
    return false;
}

bool hasAllRequiredFields(map<string, string> &passport)
{
    vector<string> required = {"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"};
    for (string &key : required)
    {
        if (passport[key] == "")
            return false;
    }
    return true;
}

bool hasAllRequiredFieldsAndValid(map<string, string> &passport)
{
    return (validateByr(passport) && validateIyr(passport) && validateEyr(passport) && validateHgt(passport) && validateHcl(passport) && validateEcl(passport) && validatePid(passport));
}

int solvePart1(vector<map<string, string> > &passports)
{
    int validCount = 0;
    for (int i = 0; i < passports.size(); i++)
    {
        map<string, string> passport = passports[i];
        if (hasAllRequiredFields(passport))
            validCount++;
    }
    return validCount;
}

int solvePart2(vector<map<string, string> > &passports)
{
    int validCount = 0;
    for (int i = 0; i < passports.size(); i++)
    {
        map<string, string> passport = passports[i];
        if (hasAllRequiredFieldsAndValid(passport))
            validCount++;
    }
    return validCount;
}

int main()
{
    vector<map<string, string> > passports = parseInput("./day4.txt");
    cout << solvePart1(passports) << endl;
    cout << solvePart2(passports) << endl;
}
